tests/dist.rs

```diff
@@ -889,17 +895,194 @@ fn removed_component() {
+    let edit = &|date: &str, chan: &mut MockChannel| {
+        if date == "2016-02-01" {
+            let tpkg = chan.packages[0]
+                .targets
+                .iter_mut()
+                .find(|p| p.target == "x86_64-apple-darwin")
+                .unwrap();
+            tpkg.components.push(MockComponent {
+                name: "bonus".to_string(),
+                target: "x86_64-apple-darwin".to_string(),
+                is_extension: true,
+            });
+            tpkg.components.push(MockComponent {
+                name: "bonus-2".to_string(),
+                target: "x86_64-apple-darwin".to_string(),
+                is_extension: true,
+            });
+        } else {
+            chan.packages
+                .retain(|p| !["bonus", "bonus-2"].contains(&p.name));
+        }
+    };
+
+    setup(
+        Some(edit),
+        false,
+        &|url, toolchain, prefix, download_cfg, temp_cfg| {
+            let adds = [
+                Component::new(
+                    "bonus".to_string(),
+                    Some(TargetTriple::new("x86_64-apple-darwin")),
+                    true,
+                ),
+                Component::new(
+                    "bonus-2".to_string(),
+                    Some(TargetTriple::new("x86_64-apple-darwin")),
+                    true,
+                ),
+            ];
+
+            // Update with bonus and bonus-2
+            change_channel_date(url, "nightly", "2016-02-01");
+            update_from_dist(
+                url,
+                toolchain,
+                prefix,
+                &adds,
+                &[],
+                &download_cfg,
+                temp_cfg,
+                false,
+            )
+            .unwrap();
+            assert!(utils::path_exists(&prefix.path().join("bin/bonus")));
+            assert!(utils::path_exists(&prefix.path().join("bin/bonus-2")));
+
+            // Update without bonus, should fail with RequestedComponentsUnavailable
+            change_channel_date(url, "nightly", "2016-02-02");
+            let err = update_from_dist(
+                url,
+                toolchain,
+                prefix,
+                &[],
+                &[],
+                &download_cfg,
+                temp_cfg,
+                false,
             )
-            .is_err());
+            .unwrap_err();
+            match *err.kind() {
+                ErrorKind::RequestedComponentsUnavailable(..) => {
+                    assert!(err.to_string().contains("rustup component remove --toolchain nightly --target x86_64-apple-darwin bonus"));
+                    assert!(err.to_string().contains("rustup component remove --toolchain nightly --target x86_64-apple-darwin bonus-2"));
+                }
+                _ => panic!(),
+            }
+        },
+    );
+}
+
+#[test]
+fn removed_components_with_different_target() {
```

Comment:
I cannot get this test to pass. Something is wrong its the setup.

Reply:
Modifying the channel is not the best way to do this.  It's better to have configured a scenario which does this kind of thing.  Ideally for comprehensibility it ought to do it with `rust-std` since that's the commonly used component which might be installed with a target different to that of the toolchain's host.

Reply:
I based the new tests on the existing `removed_component` test. I'm not really sure on how to configure a scenario for tests.